### PR TITLE
PART V: clean up mayaUsdPlugin

### PR DESCRIPTION
**Pull request by HamedSabri-adsk**
- mayaUsdPlugin promotion headers are now created under mayaUsdPlugin and not mayaUsd
- clean up include order and directive

### CODE REVIEW

File: plugin/adsk/plugin/ProxyShape.cpp [hunk 1]
```diff
@@ -13,10 +13,10 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
-#include "ProxyShape.h"
+#include <mayaUsdPlugin/ProxyShape.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/ProxyShape.cpp:16]: https://github.com/Autodesk/maya-usd/pull/390#discussion_r400242549

File: plugin/adsk/plugin/ProxyShape.h [hunk 2]
```diff
@@ -15,11 +15,12 @@
 //
 #pragma once
 
-#include "base/api.h"
+#include <mayaUsdPlugin/base/api.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/ProxyShape.h:18]: This is a private header, and should be "base/api.h" (and moved to end, regardless)

File: plugin/adsk/plugin/CMakeLists.txt [hunk 3]
```diff
@@ -93,6 +87,16 @@ if(IS_MACOSX OR IS_LINUX)
...
+# -----------------------------------------------------------------------------
+# promoted headers
+# -----------------------------------------------------------------------------
+mayaUsd_promoteHeaderList( 
+    HEADERS
+        ${headers}
+    BASEDIR
+        ${TARGET_NAME}
+)
+
```
Comment (elrond79) [inline plugin/adsk/plugin/CMakeLists.txt:99]: I don't think these need to be promoted, or should, be promoted - see my comments overall review comments.

File: plugin/adsk/plugin/ProxyShape.h [hunk 4]
```diff
@@ -15,11 +15,12 @@
...
 
+#include <mayaUsd/base/api.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/ProxyShape.h:20]: What requires this header? A quick search of the directory didn't show anything using `MAYAUSD_CORE_*`, `MAYAUSD_MACROS_*` or `MAYAUSD_TEMPLATE_*`.

File: plugin/adsk/plugin/exportTranslator.cpp [hunk 5]
```diff
@@ -14,7 +14,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
-#include "exportTranslator.h"
+#include <mayaUsdPlugin/exportTranslator.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/exportTranslator.cpp:17]: https://github.com/Autodesk/maya-usd/pull/390#discussion_r400242549

File: plugin/adsk/plugin/exportTranslator.h [hunk 6]
```diff
@@ -16,16 +16,18 @@
 //
 #pragma once
 
-#include "base/api.h"
-#include <mayaUsd/fileio/jobs/jobArgs.h>
+#include <mayaUsdPlugin/base/api.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/exportTranslator.h:19]: Think this should be left as a private header.

File: plugin/adsk/plugin/exportTranslator.h [hunk 7]
```diff
@@ -16,16 +16,18 @@
...
 
-#include "pxr/pxr.h"
+#include <mayaUsd/base/api.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/exportTranslator.h:21]: Don't see what needs this

File: plugin/adsk/plugin/exportTranslator.h [hunk 8]
```diff
@@ -16,16 +16,18 @@
...
+#include <mayaUsd/fileio/jobs/jobArgs.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/exportTranslator.h:22]: I actually consider this **7. Your project’s headers** - I was considering pretty much anything in maya-usd as in "this project" - but I'm guessing you're considering this **4. Other libraries’ headers**?

We should probably come to a decision about this - from a practical standpoint, I strongly prefer going with the anything-in-maya-usd interpretation, since that can be easily enforced via a single .clang-format file, while doing different things in different sub-directories will be much trickier to implement.

Plus, I like the simplicity of it.  And, it's more thematically in keeping with the "more general to more specific" ordering - things within the same project are more specific than general maya or pixar headers.

File: plugin/adsk/plugin/importTranslator.cpp [hunk 9]
```diff
@@ -14,28 +14,26 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
-#include "importTranslator.h"
+#include <mayaUsdPlugin/importTranslator.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/importTranslator.cpp:17]: https://github.com/Autodesk/maya-usd/pull/390#discussion_r400242549

File: plugin/adsk/plugin/importTranslator.h [hunk 10]
```diff
@@ -17,17 +17,19 @@
 #ifndef PXRUSDMAYA_IMPORT_TRANSLATOR_H
 #define PXRUSDMAYA_IMPORT_TRANSLATOR_H
 
-#include "base/api.h"
-#include <mayaUsd/fileio/jobs/jobArgs.h>
+#include <mayaUsdPlugin/base/api.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/importTranslator.h:20]: I think this should be left as a private header.

File: plugin/adsk/plugin/importTranslator.cpp [hunk 11]
```diff
@@ -14,28 +14,26 @@
...
+#include <string>
 
 #include <mayaUsd/fileio/jobs/jobArgs.h>
 #include <mayaUsd/fileio/jobs/readJob.h>
-#include <mayaUsd/fileio/shading/shadingModeRegistry.h>
 #include <mayaUsd/fileio/jobs/writeJob.h>
-
-#include "pxr/base/gf/interval.h"
-#include "pxr/base/vt/dictionary.h"
+#include <mayaUsd/fileio/shading/shadingModeRegistry.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/importTranslator.cpp:25]: Again, I consider these to be group 7.

File: plugin/adsk/plugin/importTranslator.h [hunk 12]
```diff
@@ -17,17 +17,19 @@
...
 
-#include "pxr/pxr.h"
+#include <string>
+
+#include <mayaUsd/base/api.h>
+#include <mayaUsd/fileio/jobs/jobArgs.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/importTranslator.h:25]: Group 7

File: plugin/adsk/plugin/plugin.cpp [hunk 13]
```diff
@@ -13,36 +13,38 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
+#include <mayaUsdPlugin/base/api.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/plugin.cpp:16]: I think this should be left as a private header.

File: plugin/adsk/plugin/plugin.cpp [hunk 14]
```diff
@@ -13,36 +13,38 @@
...
 
-#include "base/api.h"
-#include "importTranslator.h"
-#include "exportTranslator.h"
-#include "ProxyShape.h"
+#include <mayaUsd/base/api.h>
+
+#include <mayaUsdPlugin/exportTranslator.h>
+#include <mayaUsdPlugin/importTranslator.h>
+#include <mayaUsdPlugin/ProxyShape.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/plugin.cpp:22]: I think these should also be left as private headers.

File: plugin/adsk/plugin/plugin.cpp [hunk 15]
```diff
@@ -13,36 +13,38 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
+#include <mayaUsdPlugin/base/api.h>
 
-#include "base/api.h"
-#include "importTranslator.h"
-#include "exportTranslator.h"
-#include "ProxyShape.h"
+#include <mayaUsd/base/api.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/plugin.cpp:18]: Needed?

File: plugin/adsk/plugin/plugin.cpp [hunk 16]
```diff
@@ -13,36 +13,38 @@
...
+#include <mayaUsdPlugin/importTranslator.h>
+#include <mayaUsdPlugin/ProxyShape.h>
 
-#include <mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.h>
 #include <mayaUsd/nodes/proxyShapeBase.h>
-#include <mayaUsd/nodes/stageData.h>
 #include <mayaUsd/nodes/proxyShapePlugin.h>
+#include <mayaUsd/nodes/stageData.h>
 #include <mayaUsd/render/pxrUsdMayaGL/proxyShapeUI.h>
+#include <mayaUsd/render/vp2RenderDelegate/proxyRenderDelegate.h>
```
Comment (elrond79) [inline plugin/adsk/plugin/plugin.cpp:28]: Group 7

File: plugin/adsk/plugin/plugin.cpp [hunk 17]
```diff
@@ -13,36 +13,38 @@
...
 
 #include <mayaUsd/utils/undoHelperCommand.h>
 #if defined(WANT_QT_BUILD)
-#include <mayaUsd/ui/cmds/USDImportDialogCmd.h>
+#include <mayaUsdUI/ui/cmds/USDImportDialogCmd.h>
 #endif
 
 #if defined(WANT_UFE_BUILD)
 #include <mayaUsd/ufe/Global.h>
 #endif
```
Comment (elrond79) [inline plugin/adsk/plugin/plugin.cpp:37]: Conditional includes at end

File: plugin/adsk/plugin/plugin.cpp [hunk 18]
```diff
@@ -13,36 +13,38 @@
...
 
+#include <sstream>
```
Comment (elrond79) [inline plugin/adsk/plugin/plugin.cpp:39]: 3. C++ standard library headers - at top

**Comment by HamedSabri-adsk** [comment]
I read through all of the comments in all 5 PRs and hopefully I haven't missed anything.

Thank you @elrond79 @mattyjams @ppt-adsk @seando-adsk   again for all the great and detailed feedback. 

**Header order, blank spaces**

My original goal in this large PR was to eliminate things that clang-format couldn't handle (eliminating relative paths) and I honestly spent little on cosmetics and let clang-format later catch them. I do also understand that we have cases where clang-format can't be smart enough to handle certain ordering.  

If people are Ok, I can spend a bit more time on some of the cosmetic ordering so at least it is as close as possible to the coding guideline since I am have some changes to make in this area anyways. Any thoughts?

**Associated headers**

I am glad we all have the same understanding/agreement. Internally we have an action item to clarify this a bit more in the coding guideline document.

**Header promotion**

I do agree with all the comments around not promoting private headers. One of the main reasons we have the private header files promoted is just so that they simply end up in the build tree and help including them using the **full path** inside their own project (both in associated cpp or any other files). 

```
#include "mayaUsd/render/vp2RenderDelegate/draw_item.h"
vs simply
#include "draw_item.h"
```
To sum it up, we have 2 solutions:

A- don't promote private headers ( headers that don't installed ) and include them using **quotes** and **no full-path** which would also simplifies the CMakelists.txt
e.g 
```
draw_item.cpp
#include "draw_item.h"

basisCurves.cpp
#include "draw_item.h"
```
B- promote private headers ( headers that don't installed ) and include them using **quotes** and **full-path**

`#include "mayaUsd/render/vp2RenderDelegate/draw_item.h"`

Any Thoughts?

**Comment by kxl-adsk** [comment]
Re `Header promotion` - applying guidelines is always a great test and I agree we should reword the guideline for private header files to be more explicit about `no full-path`. Let's do this once we merge these 5 big PRs.

For your question, option A is aligned with our guidelines.

**Comment by elrond79** [comment]
Sounds good - made a PR to clarify guideline: https://github.com/Autodesk/maya-usd/pull/400
